### PR TITLE
Add connection limit history display to Stats tab

### PR DESCRIPTION
**Pull request by i-am-marvin**
Fixes #445

Renames the log viewer's Models tab to Stats and visualizes
EvalStats.connection_limit_history (adaptive-connections controller
scale changes) per the design handoff in #445:

- CONNECTIONS lane in the usage table: start/peak/avg/final stats
  (time-weighted avg), rate-limit count, and an SVG step chart with a
  time axis, configured-max guide, and rate-limit backoff dots
- Per-model change-log modal (Scoring Detail pattern) with time,
  old/new limit, and reason badges
- Narrow (<1000px) layout moves the lane to a full-width strip row
  under its model's row
- Roles view resolves lanes through the role → model alias map
- Lane renders during live evals; hidden for empty/legacy histories

Screenshot 2026-07-20 at 7 49 25 PM

Screenshot 2026-07-20 at 7 49 32 PM

Co-authored-by: Charles Teague <261654+dragonstyle@users.noreply.github.com>
Co-Authored-By: Claude Fable 5 <noreply@anthropic.com>

### CODE REVIEW

**Comment by i-am-marvin** [comment]
**Claude finished @dragonstyle's task in 19m 10s** —— [View job](https://github.com/meridianlabs-ai/ts-mono/actions/runs/29778283722)

---
### ✅ CI is green — PR ready for review

- [x] Gather context (`@auto` → verify/finish PR #447)
- [x] Sync branch with `origin/main` — already up to date (main is the branch's parent commit, no merge needed)
- [x] Review the PR diff for correctness issues — none found
- [x] Diagnose CI failure — only the `format` job failed; prettier flagged `ConnectionsLane.tsx`, `ModelTokenTable.tsx`, and `usage/index.ts`
- [x] Fix formatting and push (`7056f2f`, `7b541ad`)
- [x] Confirm CI green on the final commit — [all jobs pass](https://github.com/meridianlabs-ai/ts-mono/actions/runs/29779253133) (format, build, typecheck, lint, test, e2e)
- [x] Posted `@review` to re-run the automated reviewer

### Summary

The PR's original commit passed build/typecheck/lint/test/e2e but failed the prettier format check on three files. Since `pnpm install` isn't permitted in this environment, I couldn't run prettier locally — instead I matched the expected output against identically-shaped code in files that already pass formatting, and verified via CI:

- Collapsed the multi-line `import type { ConnectionLaneData, ConnectionWindow } from "./connectionHistory";` to a single line (exactly 80 chars, so prettier keeps it on one line) in `ConnectionsLane.tsx` and `ModelTokenTable.tsx`, and the equivalent `export type` in `usage/index.ts`.
- No functional changes; the connection-limit visualization code itself needed no fixes.
 · [Branch](https://github.com/meridianlabs-ai/ts-mono/tree/claude/issue-445-20260720-2036)

**Comment by i-am-marvin** [comment]
**Claude finished @dragonstyle's task in 17m 55s** —— [View job](https://github.com/meridianlabs-ai/ts-mono/actions/runs/29778461070)

---
### Formatting fixed & code review complete

- [x] Sync branch with `origin/main` (already up to date)
- [x] Fix `format` CI failure — prettier issues in `ConnectionsLane.tsx`, `ModelTokenTable.tsx`, `usage/index.ts`
- [x] Full CI green on `7b541ad` (format, build, typecheck, lint, test, e2e ✅)
- [x] Thorough code review (below)

**Note on formatting:** `pnpm install`/`npx` aren't permitted in this runner, so the fixes were derived from the prettier config and validated against CI. A concurrent job on this PR pushed byte-identical fixes moments before each of my pushes, so the branch history shows its commits (`7056f2f`, `7b541ad`) rather than duplicates from this run. One subtlety worth recording: prettier keeps `type CategoryKey =` followed by the union on a single indented continuation line — expanding it to leading-pipe style actually *fails* `format:check`, which is why an intermediate commit regressed before `7b541ad` restored it.

---

### Code review

Overall this is a well-built PR — no blocking bugs found. The pure helpers in `connectionHistory.ts` are cleanly separated and well tested, and the layout bookkeeping is correct. Findings, most significant first:

**1. Change-log modal state persists in the shared store and can re-open unexpectedly** — `packages/inspect-components/src/usage/UsagePanel.tsx:95`
`useProperty("usage-connections", "log-model", ...)` stores the open modal's model under a fixed id, so it survives unmount. If a user opens the change log and then navigates to a different log that has a lane for the same model name (very common when comparing runs of the same task), the modal re-opens over the new log's data. Transient per-view state elsewhere is keyed by a unique node id (e.g. `ModelEventView.tsx:50` uses `eventNode.id`). If persistence isn't deliberate here, plain `useState<string | null>` is simpler and avoids the surprise. [Fix this →](https://claude.ai/code?q=In%20packages%2Finspect-components%2Fsrc%2Fusage%2FUsagePanel.tsx%20(branch%20claude%2Fissue-445-20260720-2036)%2C%20the%20logModel%20state%20at%20line%2095%20uses%20useProperty(%22usage-connections%22%2C%20%22log-model%22)%20which%20persists%20across%20unmounts%2C%20so%20the%20connection%20change-log%20modal%20can%20re-open%20when%20navigating%20to%20another%20log%20with%20the%20same%20model.%20Replace%20it%20with%20plain%20useState%3Cstring%20%7C%20null%3E(null)%20and%20remove%20the%20now-unused%20useProperty%20import.&repo=meridianlabs-ai/ts-mono)

**2. Lanes are rebuilt on every render** — `UsagePanel.tsx:67-76`
`connectionWindow` + `buildConnectionLanes` (grouping, sorting, weighted averaging) run on each render of the panel, and `connectionsByRow` gets a fresh identity each time. Histories are small (controller scale changes), so this is fine in practice — but a `useMemo` keyed on `connection_limit_history` / `configs_by_model` / `role_aliases` would be idiomatic and cheap insurance if this panel starts re-rendering during live evals.

**3. `window` prop shadows the browser global** — `ConnectionsLane.tsx:13,87`
It reads naturally (`window.start`/`window.end`), but any future use of the real `window` inside the component (`window.addEventListener`, `window.devicePixelRatio`, …) will silently resolve to the prop. A name like `timeWindow` would remove the trap.

**4. Series can be visually squashed when the configured max is far above the peak** — `ConnectionsLane.tsx:100`
`yMax` is driven by `configuredMax`, so with max 100 and a peak of ~20 the entire step chart occupies the bottom fifth of a 46px plot. Probably intentional (it shows headroom against the dashed guide) — just confirming that's the tradeoff you want.

**5. Hardcoded light-theme colors** — `ConnectionLogModal.module.css` (badge pastels), `ConnectionsLane.module.css` (`#3a7bd5` series, `#b04a3c` dots)
Consistent with the existing hardcoded swatch palette in `ModelTokenTable.module.css:239-255`, so no action needed now — but unlike the `--bs-*` vars used for structural colors in the same files, these won't adapt if a dark theme lands.

**6. Nit:** the last x-axis tick shows time-only even when an eval spans midnight — interior ticks handle date rollover via `prevDate` (`ConnectionsLane.tsx:76-79`), but the `last` tick (`ConnectionsLane.tsx:57-61`) doesn't get the same treatment.

### What checks out well

- The time-weighted average math is correct (I verified the 20/40/80 → 55 test case by hand) and event timestamps are properly clamped to the window; edge cases (empty history, missing window, events outside eval bounds, live evals with `completed_at: ""`) are all guarded and tested.
- Legacy logs without `connection_limit_history` render no lane — the guards in `buildConnectionLanes` and `connectionWindow` handle `undefined` even though the generated type marks the field required.
- Role-view resolution is right: lanes resolve role → model through `role_aliases`, the modal is keyed by the underlying model, and roles sharing a model intentionally share a lane.
- Narrow-layout bookkeeping is exact: `columnCount` excludes the Connections column precisely when the strip row is shown, so `colSpan` always matches the header, and the strip row's suppressed top border keeps it visually attached to its model row.
- `adaptiveMaxFromConfig` covers every schema shape (`boolean`/`number`/string shorthand/object) with fallbacks matching the `AdaptiveConcurrency` default of 100.
- Accessibility is handled: `role="img"` + `aria-label` on the SVG, labeled log button, and labeled up/down arrows in the modal.
- Keeping the tab id `kLogViewModelsTabId` while renaming the label to "Stats" preserves existing deep links — assuming that's intentional, 👍.
 · [Branch](https://github.com/meridianlabs-ai/ts-mono/tree/claude/issue-445-20260720-2036)

**Comment by i-am-marvin** [comment]
@review

**Comment by dragonstyle** [comment]
@auto

**Comment by i-am-marvin** [comment]
**Claude finished @dragonstyle's task in 9m 19s** —— [View job](https://github.com/meridianlabs-ai/ts-mono/actions/runs/29779860018)

---
### ✅ Review feedback addressed — CI green

- [x] Gather context (`@auto` → address review comments on PR #447)
- [x] Sync branch with `origin/main` — already up to date (main `cbaf44a` is the branch's parent)
- [x] Fix misaligned connections cell for rows without `model_usage` (ModelTokenTable.tsx)
- [x] Rename `window` prop to `timeWindow` in ConnectionsLane.tsx
- [x] Hoist duplicated `onShowLog` callback wiring in ModelTokenTable.tsx
- [x] Verify — full CI green on `5ad7be5` (format, build, typecheck, lint, test, e2e)
- [x] Pushed `5ad7be5`, resolved all 3 review threads, posted `@review`

### Summary

Addressed all three inline comments from the automated review in commit `5ad7be5`:

1. **Grid alignment (minor correctness):** when `showTokenColumns` is true but a row has no `model_usage` entry (reachable for a model with connection history only, e.g. all requests rate-limited), the row now renders an empty `<td colSpan={2}>` placeholder so the Connections lane stays under its own header instead of shifting into the Compose column. Also added an empty per-sample `<td>` for the same case so the row's cell count always matches the header.
2. **`window` shadowing (nit):** renamed the `ConnectionsLane` prop and the `buildTicks` parameter to `timeWindow` throughout, plus both call sites in `ModelTokenTable`.
3. **Duplicated callback wiring (nit):** hoisted `showLog` to one expression per row iteration, used by both the column and strip renders.

Notes:
- `pnpm install` isn't permitted in this runner, so verification was static + CI: [all jobs pass on `5ad7be5`](https://github.com/meridianlabs-ai/ts-mono/commit/5ad7be5) (format, build, typecheck, lint, test, e2e).
- All three review threads were resolved via GraphQL, and `@review` was posted to re-run the automated reviewer on the new commit.
 · [Branch](https://github.com/meridianlabs-ai/ts-mono/tree/claude/issue-445-20260720-2036)

**Comment by i-am-marvin** [comment]
@review

**Comment by dragonstyle** [comment]
@i-am-marvin address comments (and mark any resolved that are resolved)

**Comment by i-am-marvin** [comment]
**Claude finished @dragonstyle's task in 7m 29s** —— [View job](https://github.com/meridianlabs-ai/ts-mono/actions/runs/29781295416)

---
### ✅ Review comments addressed — all threads resolved, CI green

- [x] Gather context (unresolved review threads, current code)
- [x] Sync branch with `origin/main` — already up to date (`cbaf44a` is the branch's parent)
- [x] Fix tick-interval fallback in `ConnectionsLane.tsx` (55px spacing for long windows)
- [x] Add dark-theme overrides for badge chips and chart reds (StopReasonBadge pattern)
- [x] Commit & push (`9b3c951`)
- [x] Reply to and resolve all 5 open review threads
- [x] Confirm CI green on `9b3c951` — format, build, typecheck, lint, test, e2e ×2 all pass

### Summary

Five review threads were unresolved. Two required code changes, landed in `9b3c951`; the other three were duplicates or already fixed, so they were resolved with an explanatory reply:

1. **Tick-label overlap on long windows** (`ConnectionsLane.tsx:60`) — the `kNiceIntervals` fallback capped at 1 day, so windows longer than ~6 days at column width emitted overlapping daily ticks. The fallback now rounds up to the smallest day-multiple that keeps ticks ≥55px apart: `Math.ceil((55 * span) / width / 86400) * 86400` (as suggested in the review). *(2 threads — rounds 2 & 3)*
2. **Light-theme-only colors** — added `:global([data-bs-theme="dark"])` overrides following the `StopReasonBadge.module.css` pattern: the four badge chip fills and `.arrowDown` red in `ConnectionLogModal.module.css`, plus `.series` blue, `.rateLimits`, and `.rateLimitDot` red in `ConnectionsLane.module.css`. *(2 threads — rounds 2 & 3)*
3. **`window` prop shadowing** — already fixed in `5ad7be5` (renamed to `timeWindow`); this was a duplicate thread from an earlier review round, resolved with a note.

Notes:
- `pnpm install` isn't permitted in this runner, so verification was static + CI: all jobs pass on `9b3c951` (format, build, typecheck, lint, test, e2e scout/inspect).
- The three threads from the first review round (grid alignment, `window` rename, hoisted callback) were already resolved in the previous run.
 · [Branch](https://github.com/meridianlabs-ai/ts-mono/tree/claude/issue-445-20260720-2036)

File: packages/inspect-components/src/usage/ConnectionsLane.tsx [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+  };
+  const last: Tick = {
+    x: width,
+    label: fmtTickTime(timeWindow.end),
+    anchor: "end",
+  };
+
+  const interval =
+    kNiceIntervals.find((i) => (i / span) * width >= 55) ??
+    kNiceIntervals[kNiceIntervals.length - 1]!;
```
[the remark this answers is not in the document]
Comment (i-am-marvin) [reply packages/inspect-components/src/usage/ConnectionsLane.tsx]: Fixed in 9b3c951 — the fallback now rounds up to the smallest day-multiple that keeps ticks ≥55px apart (`Math.ceil((55 * span) / width / 86400) * 86400`), exactly as suggested.

File: packages/inspect-components/src/usage/ConnectionsLane.tsx [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+  };
+  const last: Tick = {
+    x: width,
+    label: fmtTickTime(window.end),
+    anchor: "end",
+  };
+
+  const interval =
+    kNiceIntervals.find((i) => (i / span) * width >= 55) ??
+    kNiceIntervals[kNiceIntervals.length - 1]!;
```
[the remark this answers is not in the document]
Comment (i-am-marvin) [reply packages/inspect-components/src/usage/ConnectionsLane.tsx]: Fixed in 9b3c951 — the interval fallback now derives a day-multiple from the 55px spacing rule instead of capping at 1 day, so long windows can't emit overlapping daily ticks.

File: packages/inspect-components/src/usage/ConnectionsLane.tsx [hunk 3]
```diff
@@ -0,0 +1,218 @@
+import { FC, useCallback, useState } from "react";
+
+import { useResizeObserver } from "@tsmono/react/hooks";
+
+import type { ConnectionLaneData, ConnectionWindow } from "./connectionHistory";
+import styles from "./ConnectionsLane.module.css";
+
+interface ConnectionsLaneProps {
+  data: ConnectionLaneData;
+  window: ConnectionWindow;
```
[the remark this answers is not in the document]
Comment (i-am-marvin) [reply packages/inspect-components/src/usage/ConnectionsLane.tsx]: Already addressed in 5ad7be5 — the prop (and the `buildTicks` parameter) were renamed to `timeWindow`; this thread duplicated the earlier resolved one.

File: packages/inspect-components/src/usage/ConnectionLogModal.module.css [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+.badge {
+  display: inline-block;
+  padding: 2px 8px;
+  border-radius: 999px;
+  font-size: 0.7rem;
+  font-weight: 600;
+  white-space: nowrap;
+}
+
+.badgeSlowStart {
```
[the remark this answers is not in the document]
Comment (i-am-marvin) [reply packages/inspect-components/src/usage/ConnectionLogModal.module.css:77]: Fixed in 9b3c951 — added `:global([data-bs-theme="dark"])` overrides (StopReasonBadge pattern) for the badge fills and `.arrowDown` here, and for `.rateLimits`/`.rateLimitDot`/`.series` in ConnectionsLane.module.css.

[the remark this answers is not in the document]
Comment (i-am-marvin) [reply packages/inspect-components/src/usage/ConnectionLogModal.module.css:77]: Fixed in 9b3c951 — badge chips now have dark-theme overrides following the StopReasonBadge pattern (the chart colors got dark variants too, per the follow-up round).

**Comment by dragonstyle** [comment]
@i-am-marvin 

1. We need additional margin/padding between the breakdown and connections sections.
2. when the viewport is small (.e.g. 800px), the chart wraps to the second row, but extends off the screen as if there is a min-width or something forcing the view wider than the viewport.
3. can we rotate / angle the axis time labels to avoid overlapping text.

**Comment by i-am-marvin** [comment]
**Claude finished @dragonstyle's task in 15m 30s** —— [View job](https://github.com/meridianlabs-ai/ts-mono/actions/runs/29784166417)

---
### ✅ All three layout issues fixed — CI green

- [x] Gather context (sync with `origin/main` — already up to date)
- [x] 1. Add margin/padding between Breakdown and Connections sections
- [x] 2. Fix chart overflowing viewport at small widths (~800px)
- [x] 3. Angle the x-axis time labels to avoid overlap
- [x] Commit & push (`702f6ea`)
- [x] Confirm CI green — [all jobs pass](https://github.com/meridianlabs-ai/ts-mono/actions/runs/29784636209) (format, build, typecheck, lint, test, e2e)

### Summary

All three items landed in commit `702f6ea`:

1. **Section spacing** — the Connections column previously followed the Breakdown cell with only the default `0.75rem` cell gap. The Connections header and cell now get `padding-left: 2rem` (`ModelTokenTable.module.css`), and the narrow-layout strip row gets a small `6px` top gap above the lane.

2. **Viewport overflow at small widths** — this was a layout feedback loop, which is why it behaved "as if there is a min-width": the lane's SVG gets a fixed pixel `width` measured from its container, and that fixed-width SVG then contributed to the table cell's min-content width — so once rendered, the table could never shrink below the last rendered width, and the ResizeObserver never saw a smaller size. The SVG is now absolutely positioned inside the measured chart div (`ConnectionsLane.module.css`), so it no longer participates in table layout and the table shrinks freely with the viewport; the chart div carries an explicit height.

3. **Angled tick labels** — interior x-axis labels are now rotated −30° with `textAnchor="end"` (the classic d3 angled-label style), so adjacent labels can't collide even at the 55px tick spacing (55px × sin 30° ≈ 27px of perpendicular separation vs. a 9px line height); the wide date-rollover labels that previously overlapped their neighbors are covered too. The horizontal endpoint labels (dated start, end time) are kept as range annotations, and the chart height grows dynamically to fit the descent of the widest angled label so nothing clips.

Note: `pnpm install` isn't permitted in this runner, so verification was static (prettier `trailingComma: "es5"` / 80-col conventions matched by hand) plus CI, which is fully green on `702f6ea`.
 · [Branch](https://github.com/meridianlabs-ai/ts-mono/tree/claude/issue-445-20260720-2036)

**Comment by dragonstyle** [comment]
@i-am-marvin the close button is too large - it should match the appearance of the cancel button in the edit tags dialog.

**Comment by i-am-marvin** [comment]
**Claude finished @dragonstyle's task in 6m 48s** —— [View job](https://github.com/meridianlabs-ai/ts-mono/actions/runs/29787866300)

---
### ✅ Close button resized to match the edit tags dialog — CI green

- [x] Gather context (find the edit tags dialog's cancel button styling)
- [x] Sync branch with `origin/main` — already up to date
- [x] Match the Connection Log modal's close button to that appearance
- [x] Commit & push (`938ee8f`)
- [x] Confirm CI green — [CI run passes](https://github.com/meridianlabs-ai/ts-mono/actions/runs/29787943801)

### Summary

The Close button in the connection change-log modal looked larger because it was missing the `text-size-smaller` class: the Cancel button in `EditTagsDialog.tsx` uses `clsx("btn", "btn-secondary", "text-size-smaller")`, while the modal's button was plain `"btn btn-secondary"`, so it rendered at the default Bootstrap button size.

Fixed in `938ee8f` by adding `text-size-smaller` to the Close button in `ConnectionLogModal.tsx`, so it now matches the edit tags dialog's Cancel button exactly (same classes, same footer placement).

--- · [Branch](https://github.com/meridianlabs-ai/ts-mono/tree/claude/issue-445-20260720-2036)

**Review by dragonstyle — APPROVED** [review]
